### PR TITLE
Add a PFL Channel!

### LINKED ISSUE #189: Add library channel PFL preview
Essentially a stripped back player channel (or just a separate implement) that feeds directly into the headphone PFL gain node, to play things in the sidebar.

### PR DESCRIPTION
**Pull request by mstratford**
This PR essentially implements a slimmer player to the library channel.

PR highlights:
- The player itself can play from the library list without dragging into show.
- You can right click tracks in the show plan to play them without re-loading the channel player.
- The PFL player can still output to a separate hardware output, if desired.
- The timers on each channel now update more often, stops the time skipping at the previous 1sec update rate.
- Tidies up some assumed 3 channels into a single constant.

Todo:
- [x] Make loaded track draggable in show plan (to also make dragging from library easier
- [x] Name the custom output in the settings menu.

Closes #189

### CODE REVIEW

File: src/showplanner/libraries.tsx [hunk 1]
```diff
@@ -21,6 +36,131 @@ type searchingStateEnum =
   | "results"
   | "no-results";
 
+export function LibraryColumn() {
+  const [sauce, setSauce] = useState("None");
+  const dispatch = useDispatch();
+  const { auxPlaylists, managedPlaylists, userPlaylists } = useSelector(
+    (state: RootState) => state.showplan
```
Comment (markspolakovs) [inline src/showplanner/libraries.tsx:43]: Consider more granular selector scope?

File: src/mixer/audio.ts [hunk 2]
```diff
@@ -397,10 +412,10 @@ export class AudioEngine extends ((EngineEmitter as unknown) as {
     // Player Input
 
     this.playerAnalysers = [];
-    for (let i = 0; i < 3; i++) {
+    for (let i = 0; i <= PLAYER_COUNT; i++) {
```
Comment (markspolakovs) [inline src/mixer/audio.ts:415]: Shouldn't this be < not <=?

File: src/mixer/state.ts [hunk 3]
```diff
@@ -454,7 +465,18 @@ export const load = (
     const blob = new Blob([rawData]);
     const objectUrl = URL.createObjectURL(blob);
 
-    const channelOutputId = getState().settings.channelOutputIds[player];
+    let channelOutputId = getState().settings.channelOutputIds[player];
+    // If the player setting doesn't exist, reset the settings. (Happens after player count is increased)
```
Comment (markspolakovs) [inline src/mixer/state.ts:469]: At some point we'll need to write a proper settings "upgrade / migration" system (e.g. making saving the default). Out of scope of this PR though.

File: src/mixer/state.ts [hunk 4]
```diff
@@ -454,7 +465,18 @@ export const load = (
...
+    if (!channelOutputId) {
+      channelOutputId = INTERNAL_OUTPUT_ID;
+      dispatch(
+        changeSetting({
+          key: "channelOutputIds",
+          // @ts-ignore
+          val: Array(PLAYER_COUNT).fill(channelOutputId),
```
Comment (markspolakovs) [inline src/mixer/state.ts:476]: Ooer

File: src/mixer/state.ts [hunk 5]
```diff
@@ -529,6 +555,10 @@ export const load = (
       }
     });
     playerInstance.on("finish", () => {
+      // PFL on PFL Player
+      if (player === PLAYER_PFL_ID) {
+        dispatch(setChannelPFL(player, false));
+      }
```
Comment (markspolakovs) [inline src/mixer/state.ts:561]: Eh?

**Review by markspolakovs — COMMENTED** [review]
I'm not a fan of overloading the "PFL" term # it already had another meaning.

Apart from that, code looks sensible, will give it a spin later.

File: src/mixer/state.ts (hunk 3 above)
Comment (mstratford) [reply src/mixer/state.ts:469]: Indeed, my assumption was that new variables would get auto filled, but that's not the case. I was hoping to just rename the variable and set the initial to true. Sounds like this won't be the case.

File: src/mixer/state.ts (hunk 5 above)
Comment (mstratford) [reply src/mixer/state.ts:561]: Mmm, I'll give this a better comment. Essentially the player still uses the same PFL logic to the headphones that regular channels do, but in this case, we have no PFL button, it's controlled by if the player is playing or not

File: src/mixer/state.ts (hunk 4 above)
Comment (mstratford) [reply src/mixer/state.ts:476]: Is this not the same as in python when you do [Value]*2 or whatever? If it's about the @ts-ignore, that was already there, will remove if it's not necessary.

Comment (markspolakovs) [reply src/mixer/state.ts:476]: Yeah, I’m not a fan of unnecessary ts-ignores

File: src/showplanner/libraries.tsx (hunk 1 above)
Comment (mstratford) [reply src/showplanner/libraries.tsx:43]: This was already here, from the re-organising. Will try.

**Comment by mstratford** [comment]
Closes #189

**Comment by mstratford** [comment]
> I'm not a fan of overloading the "PFL" term # it already had another meaning.
> 
> Apart from that, code looks sensible, will give it a spin later.

I've renamed "Preview Player" now :)

**Review by markspolakovs — COMMENTED** [review]
Modulo the thing I've already mentioned in Slack (red flashies), looks good, gave it a spin and haven't set it on fire yet and I see no howlers in the code (although it's a lot of changes so I didn't go into too much detail).

**Comment by mstratford** [comment]
> Modulo the thing I've already mentioned in Slack (red flashies), looks good, gave it a spin and haven't set it on fire yet and I see no howlers in the code (although it's a lot of changes so I didn't go into too much detail).

Fixed in #200, since technically that isn't a change in this PR.

**Review by markspolakovs — APPROVED** [review]

**Pull request merged**